### PR TITLE
[QT-9448]: Added adapter for ace editor

### PR DESCRIPTION
**Pull request by mayankeshmishra**
Added ace adaptor to support ace editor

### CODE REVIEW

File: examples/firepad-ace-example.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  } else {
+    ref = ref.push(); // generate unique location.
+    window.location.replace(window.location + "#" + ref.key); // add it as a hash to the URL.
+  }
+  return ref;
+};
+
+const init = function (): void {
+  // Initialize Firebase.
+  firebase.initializeApp(process.env.FIREBASE_CONFIG!);
```
Comment (chandan-bhatt) [inline examples/firepad-ace-example.ts]: can create a .env-template to give context to contributors

File: src/ace-adapter.ts [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+  UndoRedoCallbackType,
+} from "./editor-adapter";
+import { EventListenerType } from './emitter';
+import { Cursor, ICursor } from './cursor';
+import { IDisposable } from './utils';
+import { ITextOperation, TextOperation } from './text-operation';
+
+export class ACEAdapter implements IEditorAdapter {
+  ignoreChanges = false;
+  private ace: Ace.Editor;
```
Comment (chandan-bhatt) [inline src/ace-adapter.ts:18]: No consistency in class member declaration (accessibility wise).

For data members we have public members followed by private ones.
For methods we have public followed by private members and then again public methods.

File: src/ace-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+  registerRedo(redoFn: UndoRedoCallbackType): void {
+    this.ace.redo = redoFn;
+  }
+
+  invertOperation(operation: ITextOperation): ITextOperation {
+    return operation.invert(this.getValue());
+  }
+
+  getText(): string {
```
Comment (chandan-bhatt) [inline src/ace-adapter.ts]: can remove monaco specific methods from `IEditorAdapter`. Can create a separate IEditorAdapter for monaco (implementing base `IEditorAdapter`).

File: src/firepad-ace.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import { v4 as uuid } from "uuid";
+import firebase from "firebase/app";
+
+import { IDatabaseAdapter, UserIDType } from "./database-adapter";
+import { FirebaseAdapter } from "./firebase-adapter";
+import { Firepad, IFirepad, IFirepadConstructorOptions } from "./firepad";
+import * as Utils from "./utils";
+import { FirestoreAdapter } from "./firestore-adapter";
+import { Ace } from "ace-builds";
```
Comment (chandan-bhatt) [inline src/firepad-ace.ts]: improper import order

File: src/firepad-ace.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ * @param databaseRef - Firebase database Reference path.
+ * @param editor - Ace Editor instance.
+ * @param options - Firepad constructor options (optional).
+ */
+export function fromAce(
+  databaseRef: string | firebase.database.Reference,
+  editor: Ace.Editor,
+  options: Partial<IFirepadConstructorOptions> = {}
+): IFirepad {
+  return fromAceWithFirebase(databaseRef, editor, options);
```
Comment (chandan-bhatt) [inline src/firepad-ace.ts]: why do we need a separate function to return `fromAceWithFirebase` function return value?

**Review by chandan-bhatt — CHANGES_REQUESTED** [review]

File: examples/firepad-ace-example.ts (hunk 1 above)
Comment (mayankeshmishra) [reply examples/firepad-ace-example.ts]: created .env-template file

File: src/ace-adapter.ts (hunk 2 above)
Comment (mayankeshmishra) [reply src/ace-adapter.ts:18]: updated member ordering public methods -> private methods

File: src/ace-adapter.ts (hunk 3 above)
Comment (mayankeshmishra) [reply src/ace-adapter.ts]: Skipping for now as discussed. Also implemented setText and getText methods to set default text here

File: src/firepad-ace.ts (hunk 4 above)
Comment (mayankeshmishra) [reply src/firepad-ace.ts]: updated import order firebase -> third party -> local

File: src/firepad-ace.ts (hunk 5 above)
Comment (mayankeshmishra) [reply src/firepad-ace.ts]: removed fromAce function

**Review by v1krishna — APPROVED** [review]

**Pull request merged**